content/sctk-simple-image-viewer.markdown

```diff
@@ -0,0 +1,658 @@
+
+The protocol is also extensible: the list of possible interfaces as well as their associated
+requests and events are defined in XML files. There is one defining the core protocol (which
+notably includes the `wl_diplay` and `wl_registry` interfaces) that all wayland server is
+expected to implement. Additional protocol extensions can also be supported, defined in other
+protocol XML files. They appear as more globals in the registry, that are simply ingored by
+clients not supporting these extensions.
+
+With this in mind, let's start writing our app with Smithay's Client Toolkit (that I'll 
+abbreviate as SCTK from now).
+
+## A simple app: image viewer
+
+Our goal here is to build a very simple client: we will start it from command-line
+giving it an image file as argument, and it will just open a window showing this image.
+We'll also make the window resizable, and it will rescale the image appropriately.
+
+### Image loading
+
+The first part or our app is thus loading this image. This is in no way Wayand-specific,
+and we'll just use the `image` crate for that:
+
+```rust
+extern crate image;
+
+use std::env;
+
+fn main() {
+    // First of all, retrieve the path from the program arguments:
+    let path = match env::args_os().skip(1).next() {
+        Some(p) => p,
+        None => {
+            println!("USAGE: ./image_wiewer <PATH>");
+            return;
+        }
+    };
+    // now, try to open the image
+    // the image crate will take care of auto-detecting the file format
+    let image = match image::open(&path) {
+        Ok(i) => i,
+        Err(e) => {
+            println!("Failed to open image {}.", path.to_string_lossy());
+            println!("Error was: {:?}", e);
+            return;
+        }
+    };
+    // We'll need the image in RGBA for drawing it
+    let image = image.to_rgba();
+
+    // ...
+}
+
+```
+
+That was the easy part, we now have the image loaded in memory. Time to initialize the
+wayland setup.
+
+### Wayland setup
+
+First of all, we need to connect to the wayland server:
+
+```rust
+extern crate smithay-client-toolkit as sctk;
+// sctk reexports wayland-client for convenience
+use sctk::reexports::client::Display;
+
+fn main() {
+    // ...
+
+    let (display, mut event_queue) = Display::connect_to_env()
+                                        .expect("Failed to connect to the wayland server.");
+
+    // ...
+}
+```
+
+This `connect_to_env()` function tells wayland-client to try to connect to a wayland
+server according to the configuration of the environment variables. This is what most
+wayland clients are expected to do.
+
+Now, from this client, we are supposed to retrieve the `wl_registry` and process its
+events to retrieve the global list and bind the globals we need. For this, SCTK provides
+us with a first utility that does that for us: `sctk::Environment`. It takes a registry
+as argument, and wraps it, processing its events, and automatically binding the most classic
```

Comment:
takes a registry as an argument, wraps it, processes its events and automatically binds [...]